File: OpenLive-iOS-Objective-C/OpenLive/BeautyEffectTableViewController.m
//
//  BeautyEffectTableViewController.m
//  OpenLive
//
//  Created by GongYuhua on 2019/3/26.
//  Copyright © 2019 Agora. All rights reserved.
//

#import "BeautyEffectTableViewController.h"

@interface BeautyEffectTableViewController ()

@property (weak, nonatomic) IBOutlet UISegmentedControl *switcher;
@property (weak, nonatomic) IBOutlet UISlider *lighteningSlider;
@property (weak, nonatomic) IBOutlet UILabel *lighteningLabel;
@property (weak, nonatomic) IBOutlet UISegmentedControl *contrastSwitcher;
@property (weak, nonatomic) IBOutlet UISlider *smoothnessSlider;
@property (weak, nonatomic) IBOutlet UILabel *smoothnessLabel;
@property (weak, nonatomic) IBOutlet UISlider *rednessSlider;
@property (weak, nonatomic) IBOutlet UILabel *rednessLabel;
@end

@implementation BeautyEffectTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.preferredContentSize = CGSizeMake(300, 219);
    
    self.switcher.selectedSegmentIndex = self.isBeautyOn ? 0 : 1;
    
    self.lighteningSlider.value = self.lightening;
    self.lighteningLabel.text = [self displayStringOfValue:self.lightening];
    
    NSInteger index = [self indexOfLevel:self.contrast];
    self.contrastSwitcher.selectedSegmentIndex = index;
    
    self.smoothnessSlider.value = self.smoothness;
    self.smoothnessLabel.text = [self displayStringOfValue:self.smoothness];
    
    self.rednessSlider.value = self.redness;
    self.rednessLabel.text = [self displayStringOfValue:self.redness];
}

- (IBAction)doSwitched:(UISegmentedControl *)sender {
    NSInteger index = sender.selectedSegmentIndex;
    self.isBeautyOn = (index == 0);
    
    if ([self.delegate respondsToSelector:@selector(beautyEffectTableVCDidChange:)]) {
        [self.delegate beautyEffectTableVCDidChange:self];
    }
}

- (IBAction)doLighteningSliderChanged:(UISlider *)sender {
    self.lightening = sender.value;
    self.lighteningLabel.text = [self displayStringOfValue:self.lightening];
    
    if ([self.delegate respondsToSelector:@selector(beautyEffectTableVCDidChange:)]) {
        [self.delegate beautyEffectTableVCDidChange:self];
    }
}

- (IBAction)doConstrastChanged:(UISegmentedControl *)sender {
    NSInteger index = sender.selectedSegmentIndex;
    self.contrast = [self levelAtIndex:index];
    
    if ([self.delegate respondsToSelector:@selector(beautyEffectTableVCDidChange:)]) {
        [self.delegate beautyEffectTableVCDidChange:self];
    }
}

- (IBAction)doSmoothnessSliderChanged:(UISlider *)sender {
    self.smoothness = sender.value;
    self.smoothnessLabel.text = [self displayStringOfValue:self.smoothness];
    
    if ([self.delegate respondsToSelector:@selector(beautyEffectTableVCDidChange:)]) {
        [self.delegate beautyEffectTableVCDidChange:self];
    }
}

- (IBAction)doRednessSliderChanged:(UISlider *)sender {
    self.redness = sender.value;
    self.rednessLabel.text = [self displayStringOfValue:self.redness];
    
    if ([self.delegate respondsToSelector:@selector(beautyEffectTableVCDidChange:)]) {
        [self.delegate beautyEffectTableVCDidChange:self];
    }
}

- (NSString *)displayStringOfValue:(CGFloat)value {
    return [NSString stringWithFormat:@"%.1f", value];
}

- (NSInteger)indexOfLevel:(AgoraLighteningContrastLevel)level {
    switch (level) {
        case AgoraLighteningContrastLow:    return 0;
        case AgoraLighteningContrastNormal: return 1;
        case AgoraLighteningContrastHigh:   return 2;
    }
}

- (AgoraLighteningContrastLevel)levelAtIndex:(NSInteger)index {
    switch (index) {
        case 0: return AgoraLighteningContrastLow;
        case 1: return AgoraLighteningContrastNormal;
        case 2: return AgoraLighteningContrastHigh;
        default: return AgoraLighteningContrastNormal;
    }
}
@end

File: OpenLive-iOS/OpenLive/BeautyEffectTableViewController.swift
//
//  EnhancerTableViewController.swift
//  AgoraLiveBroadcast
//
//  Created by GongYuhua on 16/7/6.
//  Copyright © 2016年 Agora. All rights reserved.
//

import UIKit
import AgoraRtcEngineKit

protocol BeautyEffectTableVCDelegate: NSObjectProtocol {
    func beautyEffectTableVCDidChange(_ enhancerTableVC: BeautyEffectTableViewController)
}

extension AgoraLighteningContrastLevel {
    static func allValue() -> [AgoraLighteningContrastLevel] {
        return [.low, .normal, .high]
    }
}

class BeautyEffectTableViewController: UITableViewController {

    @IBOutlet weak var switcher: UISegmentedControl!
    @IBOutlet weak var lighteningSlider: UISlider!
    @IBOutlet weak var lighteningLabel: UILabel!
    @IBOutlet weak var contrastSwitcher: UISegmentedControl!
    @IBOutlet weak var smoothnessSlider: UISlider!
    @IBOutlet weak var smoothnessLabel: UILabel!
    @IBOutlet weak var rednessSlider: UISlider!
    @IBOutlet weak var rednessLabel: UILabel!
    
    var isBeautyOn = false
    var lightening: Float = 0
    var smoothness: Float = 0
    var redness: Float = 0
    var contrast: AgoraLighteningContrastLevel = .normal
    
    weak var delegate: BeautyEffectTableVCDelegate?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        preferredContentSize = CGSize(width: 300, height: 219)
        
        switcher.selectedSegmentIndex = isBeautyOn ? 0 : 1
        
        lighteningSlider.value = lightening
        lighteningLabel.text = lightening.displayString()
        
        if let index = AgoraLighteningContrastLevel.allValue().index(of: contrast) {
            contrastSwitcher.selectedSegmentIndex = index
        }
        
        smoothnessSlider.value = smoothness
        smoothnessLabel.text = smoothness.displayString()
        
        rednessSlider.value = redness
        rednessLabel.text = redness.displayString()
    }
    
    @IBAction func doSwitched(_ sender: UISegmentedControl) {
        let index = sender.selectedSegmentIndex
        isBeautyOn = (index == 0)
        
        delegate?.beautyEffectTableVCDidChange(self)
    }
    
    @IBAction func doLighteningSliderChanged(_ sender: UISlider) {
        lightening = sender.value
        lighteningLabel.text = lightening.displayString()
        
        delegate?.beautyEffectTableVCDidChange(self)
    }
    
    @IBAction func doConstrastChanged(_ sender: UISegmentedControl) {
        let index = sender.selectedSegmentIndex
        if index < AgoraLighteningContrastLevel.allValue().count {
            contrast = AgoraLighteningContrastLevel.allValue()[index]
        }
        
        delegate?.beautyEffectTableVCDidChange(self)
    }
    
    @IBAction func doSmoothnessSliderChanged(_ sender: UISlider) {
        smoothness = sender.value
        smoothnessLabel.text = smoothness.displayString()
        
        delegate?.beautyEffectTableVCDidChange(self)
    }
    
    @IBAction func doRednessSliderChanged(_ sender: UISlider) {
        redness = sender.value
        rednessLabel.text = redness.displayString()
        
        delegate?.beautyEffectTableVCDidChange(self)
    }
}

private extension Float {
    func displayString() -> String {
        return String(format: "%.1f", self)
    }
}
